compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -700,9 +718,76 @@ public void visit(BLangTypeConversionExpr conversionExpr, AnalyzerData data) {
         validateAnnotationAttachmentCount(conversionExpr.annAttachments);
     }
 
+    public BLangNumericLiteral constructNumericLiteralFromUnaryExpr(BLangUnaryExpr unaryExpr) {
+        BLangExpression exprInUnary = unaryExpr.expr;
+        BLangNumericLiteral numericLiteralInUnary = (BLangNumericLiteral) exprInUnary;
+        Object objectValueInUnary = numericLiteralInUnary.value;
+        String strValueInUnary = String.valueOf(numericLiteralInUnary.value);
+
+        if (OperatorKind.ADD.equals(unaryExpr.operator)) {
+            strValueInUnary = "+" + strValueInUnary;
+        } else if (OperatorKind.SUB.equals(unaryExpr.operator)) {
+            strValueInUnary = "-" + strValueInUnary;
+        }
+
+        if (objectValueInUnary instanceof Long) {
+            objectValueInUnary = Long.parseLong(strValueInUnary);
+        } else if (objectValueInUnary instanceof Double) {
+            objectValueInUnary = Double.parseDouble(strValueInUnary);
+        } else if (objectValueInUnary instanceof String) {
+            objectValueInUnary = strValueInUnary;
+        }
+
+        BLangNumericLiteral newNumericLiteral = (BLangNumericLiteral)
+                TreeBuilder.createNumericLiteralExpression();
+        newNumericLiteral.kind = ((BLangNumericLiteral) unaryExpr.expr).kind;
+        newNumericLiteral.pos = unaryExpr.pos;
+        newNumericLiteral.setBType(exprInUnary.getBType());
+        newNumericLiteral.value = objectValueInUnary;
+        newNumericLiteral.originalValue = strValueInUnary;
+        newNumericLiteral.typeChecked = unaryExpr.typeChecked;
+
+        return newNumericLiteral;
+    }
+
     @Override
     public void visit(BLangFiniteTypeNode finiteTypeNode, AnalyzerData data) {
-        finiteTypeNode.valueSpace.forEach(value -> analyzeNode(value, data));
+        boolean foundUnaryExpr = false;
+        Set<BLangExpression> newValueSpace = new LinkedHashSet<>();
+        boolean isErroredExprInFiniteType = false;
+
+        for (int i = 0; i < finiteTypeNode.valueSpace.size(); i++) {
+            BLangExpression value = finiteTypeNode.valueSpace.get(i);
```

Comment:
Please extract the logic out to a method.

Reply:
Since this is the only purpose of the function why is this necessary?